### PR TITLE
Implement donate for puppy

### PR DESCRIPTION
**Pull request by lohsuan**
## Description
- Add the Puppy Detail page
- Refactor main contract & re-deployment

### CODE REVIEW

**Review by Xanonymous-GitHub — CHANGES_REQUESTED** [review]
Seems we should adjusted for the new version of the contract.
addr: 0x711672C59393883240276df8d68de897C11e5B94

https://ropsten.etherscan.io/address/0x711672c59393883240276df8d68de897c11e5b94

**Review by Xanonymous-GitHub — COMMENTED** [review]
Seems we should handle the table overflow issue on small-width window.

截圖 2022-06-12 下午3 02 25

**Comment by tzulin** [comment]
For the table overflow issue on small-width window, how about add the code below into our css, and let the table could be scrolled horizontally on small-width window.

```css
@media (max-width: 720px) {
  table {
    width: 100%;
    display: block;
    overflow-x: auto;
    white-space: nowrap;
  }
}
```

It should look like this picture.
scroll table on small-width window

File: contracts/PuppySponsor.sol [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    /**
+     * @dev donate for a specific puppy.
+     * The transaction amount is in (wei).
+     *
+     * Requirements:
+     *
+     * - `puppyId` must exist and valid.
+     */
+    function donateForPuppy(bytes32 puppyId, string calldata message, string calldata keyword) external payable {
+        require(_isValidPuppyId(puppyId));
```
Comment (tzulin) [inline contracts/PuppySponsor.sol]: How about adding an error message for it to specify why did it failed?

```suggestion
        require(_isValidPuppyId(puppyId), "Puppy does not exist.");
```

File: contracts/PuppySponsor.sol [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    function _donate(bytes32 puppyId, string calldata message, string calldata keyword) private {
+        require(msg.value > 0, "The transaction value must be greater than zero!");
+
+        donateTransactionAmount++;
+        totalReceivedDonation += msg.value;
+
+        payable(owner()).transfer(msg.value);
+
+        DonateTransaction memory newTransaction = DonateTransaction(msg.sender, owner(), msg.value, block.timestamp, puppyId, message, keyword);
+        donateTransactions.push(newTransaction);
```
Comment (tzulin) [inline contracts/PuppySponsor.sol:108]: Why don't we write it inline?

```suggestion
        donateTransactions.push(DonateTransaction(msg.sender, owner(), msg.value, block.timestamp, puppyId, message, keyword));
```

File: contracts/PuppySponsor.sol (hunk 1 above)
Comment (Xanonymous-GitHub) [reply contracts/PuppySponsor.sol]: LGTM.

File: contracts/PuppySponsor.sol (hunk 2 above)
Comment (Xanonymous-GitHub) [reply contracts/PuppySponsor.sol:108]: In fact, I think that's fine too.
The reason why it was considered to be dismantled before was that it was a little long and looked a little protruding, so it was dismantled.

Comment (tzulin) [reply contracts/PuppySponsor.sol:108]: Oh I see, then I think it's fine to keep it.

**Review by tzulin — APPROVED** [review]
LGTM

**Comment by Xanonymous-GitHub** [comment]
How about this for the table overflow issue:
截圖 2022-06-12 下午7 45 11
截圖 2022-06-12 下午7 46 31

**Comment by tzulin** [comment]
Looks great! And I'm thinking about that maybe we don't need the RECEIVER column?

**Comment by Xanonymous-GitHub** [comment]
Oh oh, I don't think it's necessary to hear you say that, it seems that it can be removed.

File: src/components/PuppyDetail.jsx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                    <a
+                      href={`https://ropsten.etherscan.io/address/${transaction.addressTo}`}
+                      target="_blank"
+                      rel="noreferrer"
+                    >
+                      <p className="text-white text-base hover:text-blue-400">
+                        {shortenAddress(transaction.addressTo)}
+                      </p>
+                    </a>
+                  </td>
```
Comment (tzulin) [inline src/components/PuppyDetail.jsx]: Remove "Receiver" column in puppy's donation table.

```suggestion
            {['Donator', 'Amount', 'Message', 'Transaction Time'].map((v, i) => (
              <th key={i} className="border border-slate-300 py-2">
                {v}
              </th>
            ))}
          </tr>
        </thead>
        <tbody>
          {transactions &&
            transactions
              .filter((x) => x.puppyId == params.id)
              .reverse()
              .map((transaction, i) => (
                <tr key={i}>
                  <td className="border border-slate-300 text-center py-1">
                    <a
                      href={`https://ropsten.etherscan.io/address/${transaction.addressFrom}`}
                      target="_blank"
                      rel="noreferrer"
                    >
                      <p className="text-white text-base hover:text-blue-400">
                        {shortenAddress(transaction.addressFrom)}
                      </p>
                    </a>
                  </td>
```

**Comment by Xanonymous-GitHub** [comment]
It will finally looks like this:
截圖 2022-06-12 下午7 56 46

File: src/components/PuppyDetail.jsx (hunk 3 above)
Comment (Xanonymous-GitHub) [reply src/components/PuppyDetail.jsx]: Addressed in [b78f18f](https://github.com/lohsuan/puppy_sponsor/pull/7/commits/b78f18fb3daf5a701d0e36c4b31a77e4bde0d9cc).

**Review by Xanonymous-GitHub — APPROVED** [review]
LGTM, But we should re-deploy the latest version of our contract after https://github.com/lohsuan/puppy_sponsor/pull/7/commits/de4dffd2627386438da24edc1e752ba7b1bb1229.

cc. @lohsuan

**Comment by lohsuan** [comment]
It seems like it only works in dark mode... 
image

It looks like the following in the  light mode.
image

Sorry to find that after merge....
Should I restore this branch to fix it?
@Xanonymous-GitHub need your comment.

**Pull request merged**